shared/health/src/main/java/io/pravega/shared/health/impl/HealthConfigImpl.java

```diff
@@ -0,0 +1,180 @@
+
+    private HealthConfigImpl(Map<String, HealthComponent> components, Set<HealthComponent> roots, Map<String, Set<String>> relations) {
+        this.components = components;
+        this.roots = roots;
+        this.relations = relations;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+
+        private final Map<String, HealthComponent> components = new HashMap<>();
+
+        private final Map<String, Set<String>> relations = new HashMap<>();
+
+        private final Set<HealthComponent> roots = new HashSet<>();
+
+        /**
+         * Defines a new {@link HealthComponent} to be created.
+         * @param name The name of the to-be {@link HealthComponent}.
+         * @param aggregator The aggregation rule to use.
+         *
+         * @return A reference to the *same* object, with updates to its internal data-structures made.
+         */
+        public Builder define(String name, StatusAggregator aggregator) {
+            if (components.containsKey(name)) {
+                log.warn("Overwriting a pre-existing component definition -- aborting.");
+                return this;
+            }
+            HealthComponent component = new HealthComponent(name, aggregator, null);
+            roots.add(component);
+            components.put(name, component);
+            relations.put(name, new HashSet<>());
+            return this;
+        }
+
+        /**
+         * Defines that there should be a parent-child relation between a {@link HealthComponent} called 'child'
+         * and a {@link HealthComponent} called 'parent'.
+         *
+         * @param child The name of the child {@link HealthComponent}.
+         * @param parent The name of the parent {@link HealthComponent}.
+         * @return A reference to the *same* object, with updates to its internal data-structures made.
+         */
+        public Builder relation(String child, String parent) {
+            if (child.equals(parent)) {
+                log.warn("Attempting to add a reference to itself -- aborting.");
+                return this;
+            }
+            if (!components.containsKey(child) || !components.containsKey(parent)) {
+                log.warn("At least one of the components in this relation has not been defined -- aborting.");
+                return this;
+            }
+            roots.remove(components.get(child));
+            relations.get(parent).add(child);
+            return this;
+        }
+
+        /**
+         * Performs an exhaustive graph traversal to ensure (DFS) that there are no cyclic relations.
+         *
+         * @return Whether or not the specification is cycle-free.
+         * @throws Exception
+         */
+        private boolean validate() {
+            boolean cycle = false;
+            for (val component : relations.entrySet()) {
+                cycle |= validate(component.getKey(), new HashMap<>());
```

Comment:
bail out early

also, why do you pass in a new hashmap every time you invoke that method? That's wasteful